packages/react/dialog/src/Dialog.tsx

```diff
@@ -220,47 +232,104 @@ const DialogContentImpl = React.forwardRef((props, forwardedRef) => {
-            onPointerDownOutside={composeEventHandlers(onPointerDownOutside, (event) => {
-              const originalEvent = event.detail.originalEvent as MouseEvent;
-              const isRightClick =
-                originalEvent.button === 2 ||
-                (originalEvent.button === 0 && originalEvent.ctrlKey === true);
-
-              // If the event is a right-click, we shouldn't close because
-              // it is effectively as if we right-clicked the `Overlay`.
-              if (isRightClick) event.preventDefault();
-            })}
-            // When focus is trapped, a focusout event may still happen.
-            // We make sure we don't trigger our `onDismiss` in such case.
-            onFocusOutside={(event) => event.preventDefault()}
-            onDismiss={() => context.onOpenChange(false)}
-          />
-        </FocusScope>
-      </RemoveScroll>
-    </Portal>
+    <>
+      <Portal>
+        <RemoveScroll>
+          <FocusScope
+            as={Slot}
+            // we make sure we're not trapping once it's been closed
+            // (closed !== unmounted when animating out)
+            trapped={context.open}
+            onMountAutoFocus={onOpenAutoFocus}
+            onUnmountAutoFocus={onCloseAutoFocus}
+          >
+            <DismissableLayer
+              role="dialog"
+              id={context.contentId}
+              aria-modal
+              aria-describedby={ariaDescribedBy || context.descriptionId}
+              // If `aria-label` is set, ensure `aria-labelledby` is undefined as to avoid confusion.
+              // Otherwise fallback to an explicit `aria-labelledby` or the ID used in the
+              // `DialogTitle`
+              aria-labelledby={ariaLabel ? undefined : ariaLabelledBy || context.titleId}
+              aria-label={ariaLabel || undefined}
+              {...contentProps}
+              ref={composedRefs}
+              disableOutsidePointerEvents
+              onEscapeKeyDown={onEscapeKeyDown}
+              onPointerDownOutside={composeEventHandlers(onPointerDownOutside, (event) => {
+                const originalEvent = event.detail.originalEvent as MouseEvent;
+                const isRightClick =
+                  originalEvent.button === 2 ||
+                  (originalEvent.button === 0 && originalEvent.ctrlKey === true);
+
+                // If the event is a right-click, we shouldn't close because
+                // it is effectively as if we right-clicked the `Overlay`.
+                if (isRightClick) event.preventDefault();
+              })}
+              // When focus is trapped, a focusout event may still happen.
+              // We make sure we don't trigger our `onDismiss` in such case.
+              onFocusOutside={(event) => event.preventDefault()}
+              onDismiss={() => context.onOpenChange(false)}
+            />
+          </FocusScope>
+        </RemoveScroll>
+      </Portal>
+      {process.env.NODE_ENV === 'development' && (
```

Comment:
This change now eliminates the need for `Slottable` that existed in `AlertDialog` because now we know when we're in `DialogContentImpl` it's because we're mounted.